### ISSUE TITLE
catkin errors when compiling snap_cam

### ISSUE THREAD

**Issue by potaito** [PX4/snap_cam#21]
Hi
I updated to the current master (c3606dd) today and can no longer compile `snap_cam` using **catkin**. Compiling commit 58b4444ec39e384b42e76dfcffb0d97413fc55c5 still works though. The problem seems to have to do with `OpticalFlow` and `klt_feature_tracker` now being git submodules. 

Following the Readme in the `snap_cam` repo and doing only 
```
mkdir -p ~/catkin_ws/src
cd ~/catkin_ws/src
catkin_init_workspace
cd src
git clone https://github.com/ros-perception/vision_opencv
git clone https://github.com/ros-perception/image_common
git clone https://github.com/PX4/snap_cam.git
cd ..
catkin_make
```
results in the following error

> -- Building snap_cam with catkin
CMake Error at snap_cam/CMakeLists.txt:40 (add_subdirectory):
  The source directory
>
>    /home/user/git/catkin_ws/src/snap_cam/src/OpticalFlow
>
>  does not contain a CMakeLists.txt file.

which is true and makes sense. So I ran 
```
cd snap_cam
git submodule update --init --recursive
```

and if I try to compile now, catkin produces another error:

> 
> CMake Error at /opt/ros/indigo/share/catkin/cmake/catkin_package.cmake:114 (message):
>   catkin_package() 'catkin' must be listed as a buildtool dependency in the
>   package.xml
> Call Stack (most recent call first):
>   /opt/ros/indigo/share/catkin/cmake/catkin_package.cmake:100 (_catkin_package)
>   snap_cam/src/OpticalFlow/external/klt_feature_tracker/CMakeLists.txt:119 (catkin_package)
> 
> 
> -- Configuring incomplete, errors occurred!
> See also "/home/user/git/catkin_ws/build/CMakeFiles/CMakeOutput.log".
> See also "/home/user/git/catkin_ws/build/CMakeFiles/CMakeError.log".
> Invoking "cmake" failed

I went on and checked `snap_cam/src/OpticalFlow/external/klt_feature_tracker/package.xml` to see if it contains catkin as a buildtool dependency; it does. So this is not the issue...

Did anyone else run into this problem? I am able to reproduce it on the snapdragon as well as on a Desktop Ubuntu 14.04. 

Building with **cmake** is still working fine!

**Comment by ChristophTobler** [comment]
@potaito Thank you for reporting. I could reproduce this on my snapdragon but could not yet figure out why that is. Need to investigate a bit more

**Comment by potaito** [comment]
Yeah it's a weird one. All I could find is that OpticalFlow, when cloned to its own catkin_ws repository, does not compile with catkin either. There it complains about a malformed `OpticalFlow/package.xml` file on line 1, column 6. Removing the copyright header all together fixed this problem and it compiled. But this does not seem to have anything to do with this issue here...

**Comment by ChristophTobler** [comment]
There is no error when removing/commenting https://github.com/ethz-ait/klt_feature_tracker/blob/34871a67648e560c7125de5ab79404f81f9d0caa/CMakeLists.txt#L119-L124 but this can not be the solution I guess...

**Comment by foehnx** [comment]
I ran into the same problem as @potaito and the quick-fix by @ChristophTobler works for the error.
After this, cmake complained about the c++ version so I had to add `add_definitions(-std=c++11)
` to `snap_cam/src/OpticalFlow/external/klt_feature_tracker/CMakeLists.txt`.

Now I'm stuck on the following linker error:
```
/home/linaro/catkin_ws/devel/.private/camera_calibration_parsers/lib/libcamera_calibration_parsers.so: undefined reference to `vtable for YAML::BadConversion'
/home/linaro/catkin_ws/devel/.private/camera_calibration_parsers/lib/libcamera_calibration_parsers.so: undefined reference to `YAML::BadConversion::~BadConversion()'
/home/linaro/catkin_ws/devel/.private/camera_calibration_parsers/lib/libcamera_calibration_parsers.so: undefined reference to `YAML::detail::node_data::convert_to_map(std::shared_ptr<YAML::detail::memory_holder>)'
/home/linaro/catkin_ws/devel/.private/camera_calibration_parsers/lib/libcamera_calibration_parsers.so: undefined reference to `typeinfo for YAML::BadConversion'
collect2: error: ld returned 1 exit status
make[2]: *** [/home/linaro/catkin_ws/devel/.private/camera_calibration_parsers/lib/camera_calibration_parsers/convert] Error 1
make[1]: *** [CMakeFiles/convert.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make: *** [all] Error 2
```

Any suggestions? I hope this might help to solve the problem completely.

**Comment by Seanmatthews** [comment]
This temporary fix is less than ideal, and it means that you can't have any projects easily depend on snap_cam, but to get it building, comment out these lines: https://github.com/ethz-ait/klt_feature_tracker/blob/34871a67648e560c7125de5ab79404f81f9d0caa/CMakeLists.txt#L119-L124

**Comment by ChristophTobler** [comment]
@Seanmatthews Thanks. But I already suggested that two comments above ;)

**Comment by Seanmatthews** [comment]
@ChristophTobler Did you figure out anything else?

**Comment by ChristophTobler** [comment]
No, unfortunately not... seems to me like catkin is having issues with nested submodules

**Comment by bilalah21** [comment]
It still have not solve the problem for me. Even after commenting out those lines as suggested.

`-- Building klt_feature_tracker as a catkin package
-- Using these message generators: gencpp;genlisp;genpy
-- klt_feature_tracker: 0 messages, 1 services
-- Building OpticalFlow with OpenCV
-- Configuring incomplete, errors occurred!
See also "/home/linaro/catkin_ws/build/CMakeFiles/CMakeOutput.log".
See also "/home/linaro/catkin_ws/build/CMakeFiles/CMakeError.log".
make: *** [cmake_check_build_system] Error 1
Invoking "make cmake_check_build_system" failed`
